webapi/recovery.go

```diff
@@ -0,0 +1,117 @@
+package webapi
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+)
+
+// Recovery returns a middleware that recovers from any panics which occur in
+// request handlers. It logs the panic, a stack trace, and the full request
+// details. It also ensure the client receives a 500 response rather than no
+// response at all.
+func Recovery() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		defer func() {
+			if err := recover(); err != nil {
+				// Check for a broken connection, as it is not really a
```

Comment:
Not an issue but you could avoid the additional indentation here if you checked `err == nil` and handled the error case after that.